xDebug enabled and pre-configured in Developer mode

aepod opened the pull request:
### Description
This pull request adds Xdebug, by adding another container and allowing nginx to route to it if the xdebug header is present. 

https://github.com/magento/magento-cloud-docker/issues/112

This PR does a couple of things:
1) Adds a container in the docker-compose.yml for xdebug connections (fpm-xdebug)
2) Modifies the Nginx Image config to support switching based on xdebug session cookie
3) One minor config change for xdebug that allows IDEs to connect more easily

This was implemented using the method that was described by di-ivanov on the github issue.



### Fixed Issues (if relevant)
1. https://github.com/magento/magento-cloud-docker/issues/112


### Manual testing scenarios
1. To test #1 you need only include the branch, and rebuild the docker config using the --mode developer flag.
`vendor/bin/ece-docker build:compose --mode developer --sync-engine native  --with-xdebug`
This should then allow you to create the containers using docker-compose. You can connect to the container and see that xdebug is enabled using `php -v`, also confirm it is not in the fpm container.

2. To test #2 from the list above, It is a little trickier. You will have to add a docker-compose.override.yml as follows:
```
version: '2.1'
services:
  web:
    build:
      context: ./vendor/magento/magento-cloud-docker/images/nginx/1.9/

```
Then you can use `docker-compose up -d --rebuild` to rebuild the container. This is only needed until nginx is published, but to test the nginx you will need the config.
Once you have done that, you should be able to use xdebug as normal, if you have the cookie set you will see traffic pass to the xdebug container.

The following chrome extension adds the proper header:
https://chrome.google.com/webstore/detail/xdebug-helper/eadndfjplgieldjbigjakmdgkmoaaaoc

Please let me know if you have any questions, I know this is a bit harder than normal to test.


### Contribution checklist
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages

aepod committed:
Added pass for xdebug requests

Dimitar Ivanov provided the code for this in the community engineering slack.

Commit:
Modified check to look for cookie, not contents. Added to v4 and v6

aepod committed:
Merge pull request #1 from magento/develop

Merging upstream into local.

Commit:
Added xdebug containers to build process for development mode only, container comes up properly with xdebug

Commit:
Added nginx switching, checks to see if XDEBUG_SESSION is set.

Commit:
Added remote_connect_back to xdebug, this helps some IDEs connect to xdebug more easily. See https://xdebug.org/docs/remote for more details.

Commit:
Exposed port 9001 in docker-compose.yml for xdebug container, this is required for connecting via xdebug

aepod commented:
I am least happy with `DeveloperBuilder` class, and maybe that one could use some refinement before accepting this. I had to snipe some code from other places, and this definitely made this class a lot heavier.  I also copy/pasted a function from another class, I was in a hurry. We can refactor that stuff out.

I just wanted to get this in  so it can be tested, and hopefully used sooner than later.

YPyltiai commented:
Hi @aepod ,

Thanks for your contribution! We will review the PR and will get back to you shortly.

Thanks,
Yevhenii

aepod committed:
Merge pull request #4 from magento/develop

MAGECLOUD-3598: Set Varnish as default cache (#97)

joeshelton-wagento commented on src/Compose/ProductionBuilder.php:
```diff
@@ -444,7 +444,7 @@ private function getVariables(Repository $config): array
+            'XDEBUG_CONFIG' => 'remote_host=host.docker.internal remote_connect_back=1',
         ];
 
         if ($this->hasSelenium($config)) {
```
I would make "serverName" and "remote_host" available to be changed via environment variable from docker-comose.yaml. People are accustomed to changing them, and sometimes it is necessary for OS compatibility reasons.

aepod replied:
This is simply the default values, developers will still be able to override this, using docker-compose.override.yml. However, I think it would have to be changed in other spots as well. 

Anyhow the change here, is just adding the remote_connection so that xdebug will work when host is not the same as the user.

joeshelton-wagento commented on src/Service/ServiceFactory.php:
```diff
@@ -54,6 +55,17 @@ class ServiceFactory
+                    '9001:9001',
+                ]
+            ]
+        ],
```
Not sure how to interpret the port configuration... Does this mean 9000 or 9001 should be listened to?

aepod replied:
FPM listens on 9000, this is only done on the internal interface in docker. Xdebug listens on 9001 and this is exported to the docker host on port 9001. This is required for the xdebug connection to work.

joeshelton-wagento replied:
I'll just comment that other ports exposed to the host use a random non-standard port, per Docker behavior when no host-facing port is specified. I strongly prefer the use of standard, well-known ports.

aepod replied:
Actually, the varnish and TLS containers expose ports this same way. 9001 is standard xdebug port. Docs say 9000 https://xdebug.org/docs/remote
Hmm, maybe I have always had it on 9001 incorrectly. Either way it has to be exposed and documented so people can use xdebug. Let me know if we should change this?

joeshelton-wagento replied:
Totally agree that keeping port related things more static is preferable.

When PHP FPM rose in popularity, many xDebug guides started recommending the use of port 9001. So, you find a lot of documentation out there recommending both 9000 and 9001. Best to just keep using the same port as before, or else it's a bit of a breaking change.

Commit:
Removed un-needed reference in the docker image generate command

Commit:
Refactored, added --with-xdebug and moved this to ProductionBuilder. This allowed a much reduced signature for DeveloperBuilder.

Commit:
Had to pull in head, should not cause any conflicts

aepod commented:
I refactored.
1) Removed the xdebug from the DeveloperBuilder, this just added a lot of bloat to that class when it was not needed.
2) Moved into ProductionBuilder, this reduced everything nicely.
3) Added a flag --with-xdebug this will allow it to run in production mode
4) Fixed one or  two other minor things related to the earlier PR
5) Merged in head, so no more merge conflict.

NadiyaS commented on src/Compose/DeveloperBuilder.php:
```diff
@@ -32,15 +33,35 @@ class DeveloperBuilder implements BuilderInterface
+    /**
+     * @var ExtensionResolver
+     */
+    private $phpExtension;
```
Remove not user property. 
The same for $serviceConfig, $converter, ...

aepod replied:
This was fixed in the refactoring, however I missed the typehints. Got them now.

NadiyaS commented on src/Compose/DeveloperBuilder.php:
```diff
@@ -10,6 +10,7 @@
 use Illuminate\Contracts\Config\Repository;
 use Magento\CloudDocker\Filesystem\FileList;
 
+
```
Remove redundant lines.
Same for line 122, 171, 172

aepod replied:
Removed extra lines.

Commit:
Applied fixes from review.

Commit:
Fixed linting (phpcs) issues

aepod commented:
I addressed the issues that phpcs was having in the travis job:


```
$ vendor/bin/phpcs ./src --standard=./tests/static/phpcs-ruleset.xml -p -n
.................................... 36 / 36 (100%)
```

aepod committed:
Merge branch 'develop' into MCD-112

aepod commented:
Was out of date with develop, so I brought it up to date.

aepod committed:
Merge branch 'develop' into MCD-112

oshmyheliuk commented on src/Compose/ProductionBuilder.php:
```diff
@@ -322,6 +322,27 @@ public function build(Repository $config): Manager
+        /**
+         * Include Xdebug if --with-xdebug is set, or if it is in developer mode.
+         */
+        if ($config->get(self::KEY_WITH_XDEBUG, false) || $config->get("mode") == "developer") {
```
Please use a constant for "mode" which you've added in the BuilderInterface

aepod replied:
We are removing this check, for consistency per Nadia and Yev's feedback.  This means this constant is not needed.

NadiyaS commented on src/Compose/BuilderInterface.php:
```diff
@@ -19,15 +19,18 @@ interface BuilderInterface
     public const SERVICE_GENERIC = 'generic';
     public const SERVICE_DB = 'db';
     public const SERVICE_FPM = 'fpm';
+    public const SERVICE_FPM_XDEBUG = 'xdebug';
```
I think it will be more clear to name it like `fpm_xdebug`

aepod replied:
Renamed this, and fixed the nginx config with it.

NadiyaS commented:
Hi @aepod, 
we apologize for delay.
Here is one thought. Why not to have the same default behavior for both modes? adding xdebug by option or having it by default as for developer as for production? 
There are 2 more comments, all other stuff is OK.

aepod commented:
Thanks for the feedback, no big deal on the delay. I will be addressing the things brought up today.

As far as why we don't want to include it in all modes, mostly just to keep docker containers to a minimum if they are not needed. Some users are on smaller laptops and having containers that they won't use may challenge limited resources.

YPyltiai commented:
@aepod , would it eat much resources if that parallel php with xdebug container is just sitting there? I also feel that different setups for these modes may add some confusion and complexity.

aepod commented:
You would never run this in production though, if someone ends up using
this for a REAL production site (it will happen if it hasn't already)...
this would be bad to have in place, very bad. It almost should be implicit
for production mode just because of that.

I defer to you, let me know which you would prefer, I will be working on
this branch today.

On Thu, Feb 6, 2020 at 11:31 AM Yevhenii Pyltiai <notifications@github.com>
wrote:

> @aepod <https://github.com/aepod> , would it eat much resources if that
> parallel php with xdebug container is just sitting there? I also feel that
> different setups for these modes may add some confusion and complexity.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magento/magento-cloud-docker/pull/122?email_source=notifications&email_token=AAA6NE4G2WDMISUSGXBU2DLRBQ3N7A5CNFSM4KGYOCB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEK73O2Q#issuecomment-582989674>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA6NEYALFO3ZDZ5TLMPQVTRBQ3N7ANCNFSM4KGYOCBQ>
> .
>

YPyltiai commented:
@aepod , I was thinking it would work like https://devdocs.magento.com/cloud/howtos/debug.html#pro-debug - meaning you need to pass separate session to get you routed through xdebug-enabled container. For all the rest traffic it won't touch it at all. Thus it will be idling most of the time. Do you see any performance implications if container just sits there?

aepod commented:
No, the danger is being able to just pass the xdebug header and see ALL php
variables. For production, this is a serious concern.

All "running" containers do take memory. and the FPM is giant in docker
over 800mb, so having two of these instanced could be a real concern for
systems with less memory available to docker.

On Thu, Feb 6, 2020 at 12:24 PM Yevhenii Pyltiai <notifications@github.com>
wrote:

> @aepod <https://github.com/aepod> , I was thinking it would work like
> https://devdocs.magento.com/cloud/howtos/debug.html#pro-debug - meaning
> you need to pass separate session to get you routed through xdebug-enabled
> container. For all the rest traffic it won't touch it at all. Thus it will
> be idling most of the time. Do you see any performance implications if
> container just sits there?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magento/magento-cloud-docker/pull/122?email_source=notifications&email_token=AAA6NE6QXLJSFXXFWHJGKEDRBRBVHA5CNFSM4KGYOCB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELABZUI#issuecomment-583015633>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA6NE7BVYZ6FEMOZ6PDBPTRBRBVHANCNFSM4KGYOCBQ>
> .
>

The virtual size of our containers is about 3gb without an extra fpm container already.
Actual in memory size when it is running is MUCH MUCH smaller, so this is not really a concern. I think performance would be fine with the extra container, I just don't think we should for security reasons.

YPyltiai commented:
@aepod , makes sense. What if we just enable it with option for both modes for consistency as @NadiyaS suggested? Shouldn't be a big deal from experience perspective I guess but will keep things more aligned to not over-complicate the flows.

aepod commented:
It makes sense to me, I will make it so it is not automatically enabled in
development mode, that it just requires the flag to enable.

On Thu, Feb 6, 2020 at 12:38 PM Yevhenii Pyltiai <notifications@github.com>
wrote:

> @aepod <https://github.com/aepod> , makes sense. What if we just enable
> it with option for both modes for consistency as @NadiyaS
> <https://github.com/NadiyaS> suggested? Shouldn't be a big deal from
> experience perspective I guess but will keep things more aligned to not
> over-complicate the flows.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magento/magento-cloud-docker/pull/122?email_source=notifications&email_token=AAA6NEZEWP7U76YFMZK7TILRBRDIHA5CNFSM4KGYOCB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELADKQI#issuecomment-583021889>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA6NE4AYZQHGI77GJZWEY3RBRDIHANCNFSM4KGYOCBQ>
> .
>

YPyltiai commented:
@aepod , great, thanks!

Commit:
Modified xdebug container name, it is now fpm_xdebug

Commit:
Removed check for development mode on xdebug container

NadiyaS commented on src/Command/BuildCompose.php:
```diff
@@ -225,6 +226,12 @@ protected function configure(): void
+                self::OPTION_WITH_XDEBUG,
+                null,
+                InputOption::VALUE_NONE,
+                'Enables XDebug while in production mode. XDebug is enabled by default in development mode'
```
Please update description

NadiyaS commented on src/Compose/BuilderInterface.php:
```diff
@@ -19,15 +19,18 @@ interface BuilderInterface
     public const DEFAULT_VARNISH_VERSION = 'latest';
     public const DEFAULT_TLS_VERSION = 'latest';
 
+    public const MODE = 'mode';
```
it is not needed anymore. The same in BuildCompose.php (line 270)

Commit:
Updated per NadiyaS comments. Removed extra parameters and updated description

aepod commented:
Updated per @NadiyaS comments.

aepod committed:
Merge branch 'develop' into MCD-112

aepod commented:
Fixed merge conflict, this should be clear for merging again.

oshmyheliuk commented on src/Service/ServiceFactory.php:
```diff
@@ -54,6 +55,17 @@ class ServiceFactory
         ],
+        self::SERVICE_FPM_XDEBUG => [
+            'image' => 'magento/magento-cloud-docker-php',
+            'ports' => [9000],[9001],
```
Please check if the array of ports is described correctly.
Looks like it should be `'ports' => [9000, 9001],` instead of `'ports' => [9000],[9001],`

aepod replied:
This is very interesting. I do NOT think ports are used from this array at least here
1) It was a typo, should have been as you described.
2) BUT, it is not used at all, neither is the FPM entry, I think these were left in, and then I cargo culted it into my container, and typod the ports.

```
        self::SERVICE_FPM => [
            'image' => 'magento/magento-cloud-docker-php',
            'ports' => [9000],
            'pattern' => '%s:%s-fpm-%s',
            'config' => [
                'extends' => self::SERVICE_GENERIC
            ]
        ],
        self::SERVICE_FPM_XDEBUG => [
            'image' => 'magento/magento-cloud-docker-php',
            'ports' => [9000,9001],
            'pattern' => '%s:%s-fpm-%s',
            'config' => [
                'extends' => self::SERVICE_GENERIC,
                'ports' => [
                    '9001:9001',
                ]
            ]
        ],
```

I think we need to remove ports, not the ones from in config array but the outer ones. I will update PR. These are not present in any other container, and everything builds and works correctly without. I even generated php images etc, these are not used that I know of.

Commit:
removed unsed port references

andriyShevtsov commented:
Internal task MAGECLOUD-5297

andriyShevtsov commented:
@aepod May you please help me to figure out next case:

1.  I built nginx image locally based on your PR
```docker build -t test/nginx:1.9 images/nginx/1.9```
2. Generate compose file for develop mode and edit it to use local nginx image ```./vendor/bin/ece-docker build:compose --mode=developer```
3. Run docker up
4. Connect to container ```docker-compose run deploy bash```

I don't see that xDebug enabled when I run command `php -v`

andriyShevtsov commented:
@aepod As I see from comments, initial implementation was changed and test scenarios  from description is not actual now.  If you  don't mind, may you please correct it?

aepod commented:
The testing notes are pretty much correct. The Xdebug will ONLY be in the fpm-xdebug container, and nginx will only route to it if you have the header cookie set.
If you want to test CLI you will need to connect to that container.

I will try and elaborate on the testing to help you, but the basic gist is correct.

andriyShevtsov commented:
I see. Thank you. I will try

andriyShevtsov commented:
It works with option ```--with-xdebug``` but it's absent  in description

QA approved

shiftedreality committed:
Merge branch 'develop' into MCD-112
